### PR TITLE
Easi 2434/mac and init user table

### PR DESCRIPTION
**Pull request by StevenWadeOddball**
# EASI-2434

## Changes and Description

- Introduce a User Table
    -  Added to the user table at request initialization
    - User Info retrieved from OktaEndpoint and stored in the DB
- Refactor NDA Agreement
   - Table now references the user table instead of EUAID

## How to test this change

1. Log into the App using your OKTA EUID Account
2.  Confirm you can log in and accept the NDA. 
3. Confirm there is a record of your account in the user table
4. If you have a test account with MAC user job codes, log in and repeat the steps for MAC

## Future Improvements
- Eventually, every table can be refactored to use the user table instead of using EUAID
- The user info endpoint is currently being constructed by combining a root URL with `/userinfo`, the actual endpoint is exposed in the frontend token, but not the backend token. It would be more ideal to get the full endpoint from there.
- The test suite should eventually add a user record to the table before every test script. I skipped that for now as it is unneeded. (This is because the account is created before a graphql request is handled.

## Front End Considerations
- With this, the MAC user functionality is mostly there, but the ability to create a new model plan needs to be removed. That will currently error out on the front end.

## PR Author Review Checklist

- [ ] Met the ticket's acceptance criteria, or will meet them in a subsequent PR.
- [ ] Added or updated tests for backend resolvers or other functions as needed.
- [ ] Added or updated client tests for new components, parent components, functions, or e2e tests as necessary.
- [ ] Updated the [Postman Collection](../MINT.postman_collection.json) if necessary.

## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.

### CODE REVIEW

**Comment by patrickseguraoddball** [comment]
This works as expected and I can verify my record in `user_account` table!

File: pkg/local/authentication_middleware.go [hunk 1]
```diff
@@ -72,17 +74,25 @@ func devUserContext(ctx context.Context, authHeader string) (context.Context, er
...
-		EUAID:             strings.ToUpper(config.EUA),
+	princ := &authentication.OKTAPrincipal{
+		Username:          strings.ToUpper(config.EUA),
 		JobCodeUSER:       swag.ContainsStrings(config.JobCodes, "MINT_USER_NONPROD"),
 		JobCodeASSESSMENT: swag.ContainsStrings(config.JobCodes, "MINT_ASSESSMENT_NONPROD"),
-	}), nil
+		JobCodeMAC:        swag.ContainsStrings(config.JobCodes, "MINT MAC Users"),
+	}
+
+	userAccount, err := userhelpers.GetOrCreateUserAccount(store, princ.ID(), true, "", "", princ.JobCodeMAC) //TODO, do we need to do anything with the user? Should we pass the id around?
```
Comment (OddTomBrooks) [inline pkg/local/authentication_middleware.go:84]: Is this TODO still relevant?

File: pkg/models/nda_agreement.go [hunk 2]
```diff
@@ -1,11 +1,22 @@
...
 // NDAAgreement represents an NDAAgreement saved in the database
 type NDAAgreement struct {
-	baseStruct
-	UserID    string     `json:"userID" db:"user_id"`
+	// baseStruct
+	UserID    uuid.UUID  `json:"userID" db:"user_id"`
 	Agreed    bool       `json:"agreed" db:"agreed"`
 	AgreedDts *time.Time `json:"agreedDts" db:"agreed_dts"`
+
+	//TODO removing baseStruct for this temporarily
```
Comment (OddTomBrooks) [inline pkg/models/nda_agreement.go:16]: TODO still relevant?

File: pkg/okta/authentication_middleware.go [hunk 3]
```diff
@@ -15,14 +15,23 @@ import (
...
 
 const (
 	jobCodeUser       = "MINT_USER_NONPROD"
 	jobCodeAssessment = "MINT_ASSESSMENT_NONPROD"
+	jobCodeMAC        = "MINT MAC Users"
 )
 
-func (f MiddlewareFactory) jwt(logger *zap.Logger, authHeader string) (*jwtverifier.Jwt, error) {
+// EnhancedJwt is the JWT and the auth token
+type EnhancedJwt struct { //TODO rename this to something logical, and return it
```
Comment (OddTomBrooks) [inline pkg/okta/authentication_middleware.go:29]: TODO still relevant?

File: pkg/server/routes.go [hunk 4]
```diff
@@ -100,11 +116,11 @@ func (s *Server) routes(
...
-	// set up Feature Flagging utilities
-	ldClient, err := flags.NewLaunchDarklyClient(s.NewFlagConfig())
-	if err != nil {
-		s.logger.Fatal("Failed to create LaunchDarkly client", zap.Error(err))
-	}
+	// // set up Feature Flagging utilities
+	// ldClient, err := flags.NewLaunchDarklyClient(s.NewFlagConfig())
+	// if err != nil {
+	// 	s.logger.Fatal("Failed to create LaunchDarkly client", zap.Error(err))
+	// }
```
Comment (OddTomBrooks) [inline pkg/server/routes.go:123]: Is there any reason to keep this?

File: pkg/storage/user_accountStore.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+	if err != nil {
+		if err.Error() == "sql: no rows in result set" { //EXPECT THERE TO BE NULL results, don't treat this as an error
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	return user, nil
+	// TODO work in progress
```
Comment (OddTomBrooks) [inline pkg/storage/user_accountStore.go:40]: Is this still a WIP?

File: pkg/storage/user_accountStore.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	}
+
+	err = statement.Get(user, userAccount)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return user, nil
+	// TODO work in progress
```
Comment (OddTomBrooks) [inline pkg/storage/user_accountStore.go:63]: Is this still a WIP?

File: pkg/userhelpers/user_account_utils.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+	jsonDataFromHTTP, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := OktaAccountInfo{}
+	err = json.Unmarshal([]byte(jsonDataFromHTTP), &ret) // TODO, add type safety, handle bad data
+
+	//TODO deserialize to another struct. Do we want another
```
Comment (OddTomBrooks) [inline pkg/userhelpers/user_account_utils.go:97]: Is this TODO still relevant?

**Review by OddTomBrooks — CHANGES_REQUESTED** [review]

File: pkg/graph/resolvers/resolver_test.go [hunk 8]
```diff
@@ -24,6 +24,7 @@ type ResolverSuite struct {
 // SetupTest clears the database between each test
 func (suite *ResolverSuite) SetupTest() {
 	err := suite.testConfigs.Store.TruncateAllTablesDANGEROUS(suite.testConfigs.Logger)
+	// suite.testConfigs.Principal = getTestPrincipal(suite.testConfigs.Store, suite.testConfigs.UserInfo.EuaUserID)
```
[the remark this answers is not in the document]
Comment (StevenWadeOddball) [reply pkg/graph/resolvers/resolver_test.go]: removed old comments

File: pkg/local/authentication_middleware.go (hunk 1 above)
Comment (StevenWadeOddball) [reply pkg/local/authentication_middleware.go]: removed old comments

**Review by OddTomBrooks — APPROVED** [review]
Great work Steven!

**Pull request merged**